components/value-distribution/src/main/java/org/datacleaner/beans/valuedist/ValueDistributionAnalyzerResultReducer.java

```diff
@@ -0,0 +1,176 @@
+            } else if (partialResult instanceof GroupedValueDistributionResult) {
+                GroupedValueDistributionResult groupedResult = (GroupedValueDistributionResult) partialResult;
+
+                for (ValueCountingAnalyzerResult valueCountingResult : groupedResult.getGroupResults()) {
+                    if (valueCountingResult instanceof SingleValueDistributionResult) {
+                        SingleValueDistributionResult singleResult = (SingleValueDistributionResult) valueCountingResult;
+                        reducedResult = reduceSingleResult(annotations, highlightedColumns, reducedResult, singleResult);
+                    } else {
+                        throw new IllegalStateException("Expected "
+                                + SingleValueDistributionResult.class.getSimpleName() + ", but encountered "
+                                + valueCountingResult.getClass().getSimpleName());
+                    }
+                }
+
+            } else {
+                throw new IllegalStateException("Unsupported type of "
+                        + ValueDistributionAnalyzerResult.class.getSimpleName() + ": "
+                        + partialResult.getClass().getSimpleName());
+            }
+        }
+
+        return reducedResult;
+    }
+
+    private SingleValueDistributionResult reduceSingleResult(final Map<String, RowAnnotation> annotations,
+            final InputColumn<?>[] highlightedColumns, final SingleValueDistributionResult reducedResult,
+            final SingleValueDistributionResult singleResult) {
+        final ValueCountList reducedTopValues = reduceTopValues(reducedResult.getTopValues(),
+                singleResult.getTopValues());
+        final int reducedDistinctCount = reduceDistinctCount(reducedResult, singleResult);
+        final int reducedTotalCount = reducedResult.getTotalCount() + singleResult.getTotalCount();
+        final Collection<String> reducedUniqueValues = reduceUniqueValues(reducedResult.getUniqueValues(),
+                singleResult.getUniqueValues());
+
+        return new SingleValueDistributionResult(reducedResult.getName(), reducedTopValues, reducedUniqueValues,
+                reducedUniqueValues.size(), reducedDistinctCount, reducedTotalCount, annotations,
+                new RowAnnotationImpl(), _rowAnnotationFactory, highlightedColumns);
+    }
+
+    private ValueCountList reduceTopValues(ValueCountList topValues1, ValueCountList topValues2) {
+        if (topValues1.getValueCounts().isEmpty()) {
+            return topValues2;
+        }
+
+        ValueCountListImpl topValuesImpl1 = (ValueCountListImpl) topValues1;
+        ValueCountListImpl topValuesImpl2 = (ValueCountListImpl) topValues2;
+
+        for (ValueFrequency valueFrequency2 : topValuesImpl2.getValueCounts()) {
+            for (ValueFrequency valueFrequency1 : topValuesImpl1.getValueCounts()) {
```

Comment:
Many caveats here:

 * If there are unique values it seems that these get added to getValueCounts() as a single entry with key LabelUtils.UNIQUE_LABEL.
 * Since you are running two for-loops inside each other the performance characteristic here will be something like Big O = O^2. I think it should be doable to do it with a Big O = 2*O if you just append each value to a single new result.